packages/react-reconciler/src/ReactFiberNewContext.js

```diff
@@ -41,35 +31,35 @@ export function pushProvider(providerFiber: Fiber): void {
     );
     context._currentRenderer = rendererSigil;
   }
+  if (inProgressContexts === null) {
+    inProgressContexts = new Set([context]);
+  } else {
+    inProgressContexts.add(context);
+  }
+  const previous = context.current;
+  context.current = {
+    previous: previous,
+    fiber: providerFiber,
+  };
 }
 
 export function popProvider(providerFiber: Fiber): void {
+  const context: ReactContext<any> = providerFiber.type.context;
+  const current = context.current;
   if (__DEV__) {
-    warning(index > -1 && providerFiber === stack[index], 'Unexpected pop.');
+    warning(
+      context.current !== null && context.current.fiber === providerFiber,
+      'Unexpected pop.',
+    );
   }
-  const changedBits = changedBitsStack[index];
-  const currentValue = currentValueStack[index];
-  changedBitsStack[index] = null;
-  currentValueStack[index] = null;
-  stack[index] = null;
-  index -= 1;
-  const context: ReactContext<any> = providerFiber.type.context;
-  context.currentValue = currentValue;
-  context.changedBits = changedBits;
+  context.current = current.previous;
 }
 
 export function resetProviderStack(): void {
-  for (let i = index; i > -1; i--) {
-    const providerFiber = stack[i];
-    const context: ReactContext<any> = providerFiber.type.context;
-    context.currentValue = context.defaultValue;
-    context.changedBits = 0;
-    changedBitsStack[i] = null;
-    currentValueStack[i] = null;
-    stack[i] = null;
-    if (__DEV__) {
-      context._currentRenderer = null;
-    }
+  if (inProgressContexts !== null) {
+    inProgressContexts.forEach(context => {
+      context.current = null;
+    });
+    inProgressContexts = null;
```

Comment:
Why not just call `clear` on a single `Set` instance?

Reply:
`context.current = null` is important.

Reply:
Right, I meant after `forEach` do `inProgressContexts.clear()` instead of nullifying it and creating a new instance each time.